File: backend/app/services/click_service.py
import hashlib
import asyncpg
from app.config import settings


def _verify_sign(data: dict) -> bool:
    raw = (f"{data.get('click_trans_id')}{settings.click_service_id}"
           f"{settings.click_secret_key}{data.get('merchant_trans_id')}"
           f"{data.get('amount')}{data.get('action')}{data.get('sign_time')}")
    return hashlib.md5(raw.encode()).hexdigest() == data.get("sign_string")


async def handle_prepare(data: dict, conn: asyncpg.Connection) -> dict:
    if not _verify_sign(data):
        return {"error": -1, "error_note": "SIGN CHECK FAILED!"}
    order = await conn.fetchrow("SELECT * FROM orders WHERE id=$1", data.get("merchant_trans_id"))
    if not order:
        return {"error": -5, "error_note": "Order not found"}
    if order["payment_status"] == "paid":
        return {"error": -4, "error_note": "Already paid"}
    payment = await conn.fetchrow(
        "INSERT INTO payments(order_id,amount,method,status,external_id) VALUES($1,$2,'click','processing',$3) RETURNING *",
        order["id"], float(data.get("amount", 0)), str(data.get("click_trans_id")),
    )
    return {"click_trans_id": data["click_trans_id"], "merchant_trans_id": data["merchant_trans_id"],
            "merchant_prepare_id": payment["id"], "error": 0, "error_note": "Success"}


async def handle_complete(data: dict, conn: asyncpg.Connection) -> dict:
    from app.socket_manager import sio
    payment = await conn.fetchrow("SELECT * FROM payments WHERE id=$1", data.get("merchant_prepare_id"))
    if not payment:
        return {"error": -6, "error_note": "Transaction not found"}
    if int(data.get("error", 0)) < 0:
        await conn.execute("UPDATE payments SET status='failed' WHERE id=$1", payment["id"])
        return {"error": 0, "error_note": "cancelled"}
    await conn.execute("UPDATE payments SET status='completed',completed_at=NOW() WHERE id=$1", payment["id"])
    await conn.execute("UPDATE orders SET payment_status='paid' WHERE id=$1", payment["order_id"])
    order = await conn.fetchrow("SELECT * FROM orders WHERE id=$1", payment["order_id"])
    if order:
        ev_data = {"order_id": payment["order_id"], "amount": float(payment["amount"]), "method": "click"}
        await sio.emit("payment_confirmed", ev_data, room=f"passenger:{order['passenger_id']}")
        if order["driver_id"]:
            await sio.emit("payment_confirmed", ev_data, room=f"driver:{order['driver_id']}")
    return {"click_trans_id": data["click_trans_id"], "merchant_trans_id": data["merchant_trans_id"],
            "merchant_confirm_id": payment["id"], "error": 0, "error_note": "Success"}

File: backend/src/services/click.service.js
const crypto = require('crypto');
const db = require('../config/database');
const { getIO } = require('../config/socket');

const SERVICE_ID = process.env.CLICK_SERVICE_ID;
const SECRET = process.env.CLICK_SECRET_KEY;

function verifySign({ click_trans_id, service_id, merchant_trans_id, amount, action, sign_time, sign_string }) {
  const str = `${click_trans_id}${service_id}${SECRET}${merchant_trans_id}${amount}${action}${sign_time}`;
  return crypto.createHash('md5').update(str).digest('hex') === sign_string;
}

async function handlePrepare({ click_trans_id, merchant_trans_id, amount, sign_string, sign_time, action }) {
  if (!verifySign({ click_trans_id, service_id: SERVICE_ID, merchant_trans_id, amount, action, sign_time, sign_string })) {
    return { error: -1, error_note: 'SIGN CHECK FAILED!' };
  }
  const order = await db('orders').where({ id: merchant_trans_id }).first();
  if (!order) return { error: -5, error_note: 'Order not found' };
  if (order.payment_status === 'paid') return { error: -4, error_note: 'Already paid' };

  const [payment] = await db('payments').insert({
    order_id: order.id,
    amount,
    method: 'click',
    status: 'processing',
    external_id: click_trans_id,
  }).returning('*');

  return { click_trans_id, merchant_trans_id, merchant_prepare_id: payment.id, error: 0, error_note: 'Success' };
}

async function handleComplete({ click_trans_id, merchant_trans_id, merchant_prepare_id, error }) {
  const payment = await db('payments').where({ id: merchant_prepare_id }).first();
  if (!payment) return { error: -6, error_note: 'Transaction not found' };

  if (error < 0) {
    await db('payments').where({ id: payment.id }).update({ status: 'failed' });
    return { click_trans_id, merchant_trans_id, merchant_confirm_id: payment.id, error: 0, error_note: 'cancelled' };
  }

  await db('payments').where({ id: payment.id }).update({ status: 'completed', completed_at: db.fn.now() });
  await db('orders').where({ id: payment.order_id }).update({ payment_status: 'paid' });

  const order = await db('orders').where({ id: payment.order_id }).first();
  const io = getIO();
  io.to(`passenger:${order.passenger_id}`).to(`driver:${order.driver_id}`).emit('payment_confirmed', {
    orderId: payment.order_id,
    amount: payment.amount,
    method: 'click',
  });

  return { click_trans_id, merchant_trans_id, merchant_confirm_id: payment.id, error: 0, error_note: 'Success' };
}

module.exports = { handlePrepare, handleComplete };
